ceph-releases/kraken/ubuntu/16.04/daemon/osd_scenarios/osd_disk_prepare.sh

```diff
@@ -6,62 +6,63 @@ function osd_disk_prepare {
+      ceph-disk -v zap ${OSD_DEVICE}
     else
-      ceph-disk -v prepare ${CEPH_OPTS} --journal-uuid ${OSD_JOURNAL_UUID} ${OSD_DEVICE} ${OSD_JOURNAL}
+      log "Regarding parted, device ${OSD_DEVICE} is inconsistent/broken/weird."
+      log "It would be too dangerous to destroy it without any notification."
+      log "Please set OSD_FORCE_ZAP to '1' if you really want to zap this disk."
+      exit 1
     fi
-    chown ceph. ${OSD_JOURNAL}
-  else
-    if [[ ${OSD_BLUESTORE} -eq 1 ]]; then
-      ceph-disk -v prepare ${CEPH_OPTS} --bluestore ${OSD_DEVICE}
-    elif [[ ${OSD_DMCRYPT} -eq 1 ]]; then
-      get_admin_key
-      check_admin_key
-      # the admin key must be present on the node
-      # in order to store the encrypted key in the monitor's k/v store
-      ceph-disk -v prepare ${CEPH_OPTS} --journal-uuid ${OSD_JOURNAL_UUID} --lockbox-uuid ${OSD_LOCKBOX_UUID} --dmcrypt ${OSD_DEVICE}
-      echo "Unmounting LOCKBOX directory"
-      # NOTE(leseb): adding || true so when this bug will be fixed the entrypoint will not fail
-      # Ceph bug tracker: http://tracker.ceph.com/issues/18944
-      umount /var/lib/ceph/osd-lockbox/$(blkid -o value -s PARTUUID ${OSD_DEVICE}1) || true
+  fi
+
+  # then search for some ceph metadata on the disk
+  if [[ "$(parted --script ${OSD_DEVICE} print | egrep '^ 1.*ceph data')" ]]; then
+    if [[ ${OSD_FORCE_ZAP} -eq 1 ]]; then
+      log "It looks like ${OSD_DEVICE} is an OSD, however OSD_FORCE_ZAP is enabled so we are zapping the device anyway"
+      ceph-disk -v zap ${OSD_DEVICE}
     else
-      ceph-disk -v prepare ${CEPH_OPTS} --journal-uuid ${OSD_JOURNAL_UUID} ${OSD_DEVICE}
+      log "INFO- It looks like ${OSD_DEVICE} is an OSD, set OSD_FORCE_ZAP=1 to use this device anyway and zap its content"
+      log "You can also use the zap_device scenario on the appropriate device to zap it"
+      log "Moving on, trying to activate the OSD now."
     fi
+  fi
+
+  if [[ ${OSD_BLUESTORE} -eq 1 ]]; then
+    ceph-disk -v prepare ${CEPH_OPTS} --bluestore ${OSD_DEVICE} ${OSD_JOURNAL}
+  elif [[ ${OSD_DMCRYPT} -eq 1 ]]; then
+    get_admin_key
+    check_admin_key
+    # the admin key must be present on the node
+    # in order to store the encrypted key in the monitor's k/v store
+    ceph-disk -v prepare ${CEPH_OPTS} --journal-uuid ${OSD_JOURNAL_UUID} --lockbox-uuid ${OSD_LOCKBOX_UUID} --dmcrypt ${OSD_DEVICE} ${OSD_JOURNAL}
+    echo "Unmounting LOCKBOX directory"
+    # NOTE(leseb): adding || true so when this bug will be fixed the entrypoint will not fail
+    # Ceph bug tracker: http://tracker.ceph.com/issues/18944
+    DATA_UUID=$(blkid -o value -s PARTUUID ${OSD_DEVICE}1)
+    umount /var/lib/ceph/osd-lockbox/${DATA_UUID} || true
+  else
+    ceph-disk -v prepare ${CEPH_OPTS} --journal-uuid ${OSD_JOURNAL_UUID} ${OSD_DEVICE} ${OSD_JOURNAL}
+  fi
+
+  if [[ -n "${OSD_JOURNAL}" ]]; then
     chown ceph. $(dev_part ${OSD_DEVICE} 2)
+  else
+    chown ceph. ${OSD_JOURNAL}
```

Comment:
This `else` can not work.
`-n` is true if the length of string is nonzero.

In the case it's False, we enter the `else` but `OSD_JOURNAL` is empty so this fails.